com.ca.lsp.cobol/lsp-core-cobol-parser/src/main/java/com/ca/lsp/core/cobol/preprocessor/sub/document/impl/CobolSemanticParserListenerImpl.java

```diff
@@ -14,130 +14,297 @@
-    createVariableAndDefine(ctx);
+  public void enterReplaceOffStatement(@Nonnull ReplaceOffStatementContext ctx) {
+    cleaner.push();
+  }
+
+  @Override
+  public void enterTitleStatement(@Nonnull TitleStatementContext ctx) {
+    cleaner.push();
+  }
+
+  @Override
+  public void enterCopyStatement(@Nonnull CopyStatementContext ctx) {
+    cleaner.push();
+  }
+
+  @Override
+  public void exitCompilerOptions(@Nonnull CompilerOptionsContext ctx) {
+    // throw away COMPILER OPTIONS terminals
+    cleaner.pop();
   }
 
   @Override
   public void exitCopyStatement(@Nonnull CopyStatementContext ctx) {
+
+    // throw away COPY terminals
+    cleaner.pop();
+
+    // a new context for the copy book content
+    cleaner.peek().write(CPY_ENTER_TAG);
     /*
-     * define the copy book
+     * copy the copy book
      */
-    CopySourceContext copySource = ctx.copySource();
-    String copybookName = retrieveCopybookName(copySource);
-    Position position = retrievePosition(copySource);
-    defineCopybook(copybookName, position);
+    String copybookName = retrieveCopybookName(ctx.copySource());
+    Position position = retrievePosition(ctx.copySource());
+
+    CopybookModel model = getCopyBookContent(copybookName, position);
+    String copybookContent = model.getContent();
+
+    addCopybookUsage(copybookName, position);
+    addCopybookDefinition(copybookName, model.getUri());
+
+    List<ReplacingPhraseContext> replacingPhraseContexts = ctx.replacingPhrase();
+    if (!replacingPhraseContexts.isEmpty()) {
+      applyReplacing(copybookName, model.getUri(), copybookContent, replacingPhraseContexts);
+    } else {
+      writeCopybookContent("<URI>" + model.getUri() + "</URI>", copybookContent);
+    }
+
+    String content = cleaner.peek().read();
+    cleaner.pop();
+
+    cleaner.peek().write(content);
+  }
+
+  private void applyReplacing(
+      String copybookName,
+      String uri,
+      String copybookContent,
+      List<ReplacingPhraseContext> replacingPhraseContexts) {
+    String copybookWithReplacingName =
+        getUniqueNameForReplacing(copybookName, replacingPhraseContexts);
+
+    cleaner.peek().write("<URI>" + copybookWithReplacingName + "</URI>. ");
+    CobolDocumentContext documentContext = cleaner.push();
+    replacingPhraseContexts.forEach(
+        it -> documentContext.storeReplaceablesAndReplacements(it.replaceClause()));
+
+    documentContext.write(copybookContent);
+    documentContext.replaceReplaceablesByReplacements(tokens);
+
+    documentMappings.put(
+        copybookWithReplacingName,
+        convertTokensToPositions(uri, retrieveTokens(documentContext.read())));
+    documentContext.write(CPY_EXIT_TAG);
+  }
+
+  private String getUniqueNameForReplacing(
+      String copybookName, List<ReplacingPhraseContext> replacingPhraseContexts) {
+    return copybookName
+        + replacingPhraseContexts.stream()
+            .map(RuleContext::getText)
+            .reduce((r1, r2) -> r1 + r2)
```

Comment:
replaceable with Integer::sum

Reply:
Why? `r1` and `r2` are strings

Reply:
Sorry, I mean:  reduce("", String::concat)